stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt

```diff
@@ -0,0 +1,797 @@
+     */
+    private suspend fun RepositoryFacade.hasReadEventsCapability(cid: String): Boolean {
+        return selectChannels(listOf(cid)).let { channels ->
+            val channel = channels.firstOrNull()
+            if (channel?.ownCapabilities?.contains(ChannelCapabilities.READ_EVENTS) == true) {
+                true
+            } else {
+                logger.d {
+                    "Skipping unread counts update for channel: $cid. ${ChannelCapabilities.READ_EVENTS} capability is missing."
+                }
+                false
+            }
+        }
+    }
+
+    private fun Message.enrichWithOwnReactions(batch: EventBatchUpdate, currentUserId: UserId, eventUser: User?) {
+        ownReactions = if (eventUser != null && currentUserId != eventUser.id) {
+            batch.getCurrentMessage(id)?.ownReactions ?: mutableListOf()
+        } else {
+            mergeReactions(
+                latestReactions.filter { it.userId == currentUserId },
+                batch.getCurrentMessage(id)?.ownReactions ?: mutableListOf()
+            ).toMutableList()
+        }
+    }
+
+    private fun GlobalMutableState.updateCurrentUser(me: User) {
+        _user.value = me
+        _mutedUsers.value = me.mutes
+        _channelMutes.value = me.channelMutes
+        _totalUnreadCount.value = me.totalUnreadCount
+        _channelUnreadCount.value = me.unreadChannels
+        _banned.value = me.banned
+    }
+
+    private infix fun UserId.mustBe(currentUserId: UserId?) {
+        if (this != currentUserId) {
+            throw InputMismatchException(
+                "received connect event for user with id $this while for user configured " +
+                    "has id $currentUserId. Looks like there's a problem in the user set"
+            )
+        }
+    }
+
+    companion object {
+        val EMPTY_DISPOSABLE = object : Disposable {
+            override val isDisposed: Boolean = true
+            override fun dispose() {}
+        }
+    }
+}
+
+private typealias UserId = String
+
+private class BatchEvent(
+    val sortedEvents: List<ChatEvent>,
+    val isFromHistorySync: Boolean,
+) {
+    val size: Int get() = sortedEvents.size
+    val isFromSocketConnection get() = !isFromHistorySync
+}
+
+private class SocketEventCollector(
+    private val scope: CoroutineScope,
+    private val fireEvent: suspend (BatchEvent) -> Unit
+) {
+    private val mutex = Mutex()
+    private val postponed = arrayListOf<ChatEvent>()
+    private val timeoutJob = AtomicReference<Job>()
+
+    suspend fun add(event: ChatEvent): Boolean {
+        if (event is UserStartWatchingEvent || event is UserStopWatchingEvent) {
+            StreamLog.d(TAG) { "[add] event.type: ${event.type}" }
+            mutex.withLock {
+                return postponed.add(event).also {
+                    scheduleTimeout()
+                }
+            }
+        }
+        return false
+    }
+
+    private fun scheduleTimeout() {
+        timeoutJob.get()?.cancel()
```

Comment:
🤔 It's an edge case but what will happen if we are going to constantly receive events every 200ms?

Reply:
Good thinking.
We should probably limit the max size of the batch.